File: frontend/src/api/traffic.ts
import type {
  MetricDirection,
  TrafficMetricOut,
  WeekColumnOut,
} from '@/types/api';
import { api } from './client';

// `end` is the Monday (YYYY-MM-DD) of the newest week in the window; omit to
// let the server default to the current week. `count` = window size.
function windowQuery(end: string | null, count: number): string {
  const params = new URLSearchParams({ count: String(count) });
  if (end) params.set('end', end);
  return params.toString();
}

export function getWeekColumns(end: string | null, count: number): Promise<WeekColumnOut[]> {
  return api<WeekColumnOut[]>(`/traffic/weeks?${windowQuery(end, count)}`);
}

export function listMetrics(end: string | null, count: number): Promise<TrafficMetricOut[]> {
  return api<TrafficMetricOut[]>(`/traffic/metrics?${windowQuery(end, count)}`);
}

export interface CreateMetricInput {
  name: string;
  unit?: string | null;
  direction: MetricDirection;
  weekly_target?: number | null;
  north_star_target?: number | null;
  editor_ids?: number[];
  viewer_ids?: number[];
}

export function createMetric(input: CreateMetricInput): Promise<TrafficMetricOut[]> {
  return api<TrafficMetricOut[]>('/traffic/metrics', { method: 'POST', body: input });
}

export interface UpdateMetricInput {
  name?: string;
  unit?: string | null;
  direction?: MetricDirection;
  weekly_target?: number | null;
  north_star_target?: number | null;
  sort_order?: number;
  editor_ids?: number[];
  viewer_ids?: number[];
}

export function updateMetric(id: number, input: UpdateMetricInput): Promise<TrafficMetricOut[]> {
  return api<TrafficMetricOut[]>(`/traffic/metrics/${id}`, { method: 'PATCH', body: input });
}

export function deleteMetric(id: number): Promise<null> {
  return api<null>(`/traffic/metrics/${id}`, { method: 'DELETE' });
}

export interface UpsertValueInput {
  // status is computed server-side from value vs. weekly target — not sent by the client.
  value?: number | null;
  note?: string | null;
}

// `weekStart` is the Monday (YYYY-MM-DD) of the target week.
export function upsertMetricValue(
  assignmentId: number,
  weekStart: string,
  input: UpsertValueInput,
): Promise<TrafficMetricOut> {
  return api<TrafficMetricOut>(`/traffic/metric-assignments/${assignmentId}/values/${weekStart}`, {
    method: 'PATCH',
    body: input,
  });
}

File: backend/app/schemas/traffic.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.common import ORMModel

_DIRECTIONS = {"increase", "decrease"}


class WeekColumnOut(BaseModel):
    week_index: int  # ISO week number, used as the column label key
    label: str
    week_start: date
    week_end: date
    is_empty: bool = False


class TrafficMetricValueOut(ORMModel):
    id: int
    metric_id: int
    assignment_id: int
    week_start: date
    week_end: date
    value: Decimal | None
    status: str  # on_target | missed
    note: str | None


class TrafficMetricMemberOut(BaseModel):
    user_id: int
    name: str
    avatar_url: str | None
    role: str


class TrafficMetricAssigneeOut(BaseModel):
    assignment_id: int
    user_id: int
    name: str
    avatar_url: str | None
    effective_from: date


class TrafficMetricOut(ORMModel):
    id: int
    assignment_id: int | None = None
    owner_id: int
    assignee: TrafficMetricAssigneeOut | None = None
    name: str
    unit: str | None
    direction: str
    weekly_target: Decimal | None
    north_star_target: Decimal | None
    sort_order: int
    values: list[TrafficMetricValueOut]  # only weeks inside the current window
    recent_avg: Decimal | None  # mean of filled values in the window
    status: str  # rollup for the current window: on_target | missed | empty
    members: list[TrafficMetricMemberOut] = []
    assignees: list[TrafficMetricAssigneeOut] = []
    my_role: str = "owner"
    can_edit_values: bool = True
    can_edit_meta: bool = True
    can_manage_members: bool = True
    can_delete: bool = True
    is_pending: bool = False  # newest window week not yet filled and I can fill


def _validate_direction(value: str | None) -> str | None:
    if value is not None and value not in _DIRECTIONS:
        raise ValueError("direction must be increase/decrease")
    return value


def _validate_name(value: str | None) -> str | None:
    if value is None:
        return None
    normalized = value.strip()
    if not normalized:
        raise ValueError("name cannot be blank")
    return normalized


class TrafficMetricCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=200)
    unit: str | None = Field(default=None, max_length=30)
    direction: str = "increase"
    weekly_target: Decimal | None = None
    north_star_target: Decimal | None = None
    editor_ids: list[int] = Field(default_factory=list)
    viewer_ids: list[int] = Field(default_factory=list)

    @field_validator("direction")
    @classmethod
    def validate_direction(cls, value: str) -> str:
        return _validate_direction(value)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        return _validate_name(value)


class TrafficMetricUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=200)
    unit: str | None = Field(default=None, max_length=30)
    direction: str | None = None
    weekly_target: Decimal | None = None
    north_star_target: Decimal | None = None
    sort_order: int | None = None
    editor_ids: list[int] | None = None
    viewer_ids: list[int] | None = None

    @field_validator("direction")
    @classmethod
    def validate_direction(cls, value: str | None) -> str | None:
        return _validate_direction(value)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str | None) -> str | None:
        return _validate_name(value)


class TrafficMetricValueUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # Per-week status is computed server-side from the value vs. weekly target;
    # the client no longer picks the traffic-light color manually.
    value: Decimal | None = None
    note: str | None = None
